src/ucs/sys/sys.c

```diff
@@ -979,54 +1016,32 @@ ucs_status_t ucs_sys_enum_pfn(uintptr_t address, unsigned page_count,
         return UCS_ERR_IO_ERROR; /* could not open file */
     }
 
-    offset = (address / ucs_get_page_size()) * sizeof(data);
-    len    = sizeof(*data) * page_count;
-    data   = ucs_malloc(len, "pfn array");
-    if (data == NULL) {
-        return UCS_ERR_NO_MEMORY;
-    }
+    data = ucs_alloca(ucs_min(UCS_SYS_ENUM_PFN_ELEM_CNT, page_count) *
+                      sizeof(*data));
 
-    ret = pread(pagemap_fd, data, len, offset);
-    if (ret < 0) {
-        ucs_warn("pread(file=%s offset=%zu) failed: %m", pagemap_file, offset);
-        status = UCS_ERR_IO_ERROR;
-        goto out;
+    for (page_num = 0; (page_num < page_count) && (status == UCS_OK);
+         page_num += UCS_SYS_ENUM_PFN_ELEM_CNT) {
+         status = ucs_sys_enum_pfn_internal(pagemap_fd, page_num, data, address,
```

Comment:
adjust address?
pls add gtest which uses ucs_sys_enum_pfn() on a large region and makes sure every returned PFN is different (via std::set), this test would fail now, fix it by adjusting address by page_num*page_size

Reply:
adjusting of address is not required - there used integer division which just cut unaligned tail.
not sure about why test should fail, added gtest

Reply:
right, i see start_page is also passed.